### ISSUE TITLE
[Feature Request] Foxhound compat

### ISSUE THREAD

**Issue by Krutoy242** [Invadermonky/MagicultureIntegrations#2]
`Foxhound` entity from Quark is another "Furnace Heater" mechanic.
Right now it working only with vanilla furnaces, but it would be awesome if it work with all of the furnaces you added compat with already, such as MA Essence Furnaces.

[![](http://img.youtube.com/vi/IUAr9T6vY4g/0.jpg)](https://youtu.be/IUAr9T6vY4g?t=1376)

**Comment by Invadermonky** [comment]
This should be easy enough. The heatable system used in the mod is designed to be expandable so all I would need to do is build a handler for the entity tick.

Are you aware of any other furnaces/heaters that you would like included?

**Comment by Krutoy242** [comment]
There is a list of "furnaces" that i found in my modpack, that can be "heated up" with fox or other mechanics:

- **Seared Furnace**, **Smeltery Controller** from `TConstruct`: working on fluid fuel, but still can be speeded up.
- **Stirling Generator**, **Blast Furnace**, **Fermenter** from `IC2`: not actual furnaces, but since they cant be overclocked as other IC2 machines would be cool to have fox for that.
- **Coke Kiln** from `IC2`.
- **Melter** from `Tinkers' Complement`.  
  ![image](https://github.com/user-attachments/assets/5f54fad3-b82b-48de-b3d9-b0bf0e1b482f)
- **Blast Furnace**, **Coke Kiln**, **Arc Furnace** from `Immersive Engineering`: maybe Arc Furnace idea is bit controversal, but blast and Kiln is definitely match idea of fox butt heating up.  
  ![image](https://github.com/user-attachments/assets/4f61b908-3785-407c-a609-c8dab8aed636)
- **Small Laboratory Furnace** from `Engineer's Decor`.  
  ![image](https://github.com/user-attachments/assets/127dc0cf-72a9-4baf-b12d-e48edcb86632)
- **Essentia Smelteries** from `Thaumcraft` and `Thaumic Additions - Reconstructed`.  
  ![image](https://github.com/user-attachments/assets/176569ec-5ff3-457d-a02e-46c8860601a5)

**Comment by Invadermonky** [comment]
Multiblock structures are kind of an iffy thing to add to the heaters. While I'm not opposed to adding things like IE's Coke Oven or Blast Furnace, it is more a question of implementation.

Mutliblock structures generally have one of two possible constructions:

1. The multiblock uses a "controller" block that contains all the information and update ticking stuff (Tinker's Smeltery is a good example of this). Usually, you can only access the multiblock by interacting with this block.
2. The multiblock uses "dummy" tile entities to fill out the structure (IE and Railcraft use this method). This allows players to click on any part of the structure to open the interface.

With \# 1, there are only a couple faces you can have the heatable interaction with. This is the most balanced way to handle it, but it means that often there is only one accessible heatable face on the entire machine.

\# 2 is where things get really nasty. Since every block of the structure is a "dummy" connected to the main tile entity, mod authors will usually put the real block in the middle of the multiblock. This means to access that block with something like the Foxhound, you would have to detect the dummies then tie it into the "real" controller.

This doesn't seem too bad until you consider that the IE coke oven has 26 "dummy" blocks accounting for 54 accessible block faces. Players min/maxing would fill every face with heaters (plus more if using Phantom Faces). This could very easily destroy both performance as well as balance.

Using IE's External Heater as an example and adding one to each of the 54 faces on a Coke Oven, every second of crafting operation would result in 54 seconds of reduced crafting time.

**TLDR;** Foxhound and similar directional heaters are probably out of the question for multiblocks, but area scanning heaters like Exoflame may work.

**Comment by Krutoy242** [comment]
What I imagining is how I have huge fox farm breeding them and how my factory full of sitting foxes on every heat-required machine 😁

Some other machines I may forgot:
- **Heat Generator** from `Mekanism`
- **Alloy Tank** from `Tinkers' Complement`

![image](https://github.com/user-attachments/assets/8525e0df-0f3c-4cf4-af28-1fecda897f37)

**Comment by Invadermonky** [comment]
I was doing some testing with the Foxhound and my version is significantly slower than the one in your video. Did you make any changes to the Foxhound in your modpack?

Quark code only seems to reduce the cook time by 1 tick every 3 ticks.

**Comment by Krutoy242** [comment]
Yes, in E2EE foxhound is about 30 times more efficient, x10 speed up of furnace. [I patched Quark with Bansoukou](https://github.com/Krutoy242/Enigmatica2Expert-Extended/tree/master/dev/automation/data/bansoukou_diffs/QuarkRotN-r1.6-189/vazkii/quark/world/entity).
For your MI mod I would like to have configurable speeding up parameter, since default 33% speeding up is kinda weak. Even after i buffed Foxhounds about 2 years ago, they still doesnt become meta for players (probably because of unsitting after tamer changing dim, but this is other issue).

**Comment by Invadermonky** [comment]
Worked on this a bit and have a few updates about it.
- IC2's Sterling Generator and Mekanism's Heat Generator are both no-go for the Foxhound as both of these require TileEntities feeding heat to them. There is no easy (or non-hacky) way to accomplish this.
- Thaumic Additions starts screaming at me if I try to make changes to anything so I honestly can't be bothered trying to mess with it.

**Comment by Krutoy242** [comment]
What do you meant by "start screaming"? It have mechanics to prevent doing mixins?

**Comment by Invadermonky** [comment]
I took a look at Thaumic Additions today and I managed to get the mixins working, but loading it in my dev instance throws a lot of errors to the log from both Thaumic Additions and HammerLib.

It doesn't seem to do it on clients when the jar is built, so I'm tentatively including the integration. If it becomes a problem in the future I may remove it again.

**Comment by Invadermonky** [comment]
## Update ##
The mixin rewrite is finished and seems to be working.  I need to do some checking to make sure that nothing breaks outside of the development environment but it appears stable.

No ETA on the release as the current plan is to remove the Survival module (Tough as Nails/SimpleDifficulty) and move those features into their own mod. The release of this update will be delayed until the survival tools mod is completed.

## Requested Features ##
| Mod                   | Block                    | Type      |  Status   | Information                                                                |
|-----------------------|--------------------------|-----------|:---------:|----------------------------------------------------------------------------|
| Engineer's Decor      | Small Laboratory Furnace | Heatable  | COMPLETED |                                                                            |
| Immersive Engineering | Alloy Smelter            | Boostable | COMPLETED | Can be boosted from all accessible faces, can break balance                |
| Immersive Engineering | Arc Furnace              | Boostable |  FAILED   | Due to complexity of the Arc Furnace, adding this could break other things |
| Immersive Engineering | Blast Furnace            | Boostable | COMPLETED | Can be boosted from all accessible faces, can break balance                |
| Immersive Engineering | Coke Kiln                | Boostable | COMPLETED | Can be boosted from all accessible faces, can break balance                |
| Industrial Craft      | Blast Furnace            | Boostable | COMPLETED |                                                                            |
| Industrial Craft      | Coke Kiln                | Boostable | COMPLETED | Can only be boosted from Coke Kiln block (1 accessible face)               |
| Industrial Craft      | Fermenter                | Boostable |  FAILED   | Would have to check for each type of unique heat generator to work         |
| Industrial Craft      | Stirling Generator       |           |  FAILED   | Stirling Generator needs to attach to a tile entity to work                |
| Mekanism              | Heat Generator           |           |  FAILED   | Heat Generator needs to attach to a tile entity to work                    |
| Quark                 | Foxhound                 | Provider  | COMPLETED | Includes configurable boost bonus                                          |
| Thaumcraft            | Essentia Smelteries      | Heatable  | COMPLETED |                                                                            |
| Thaumic Additions     | Essentia Smelteries      | Heatable  | COMPLETED |                                                                            |
| Tinker's Complement   | Alloy Tank               | Boostable |  FAILED   | This is just a tank that does extra stuff, no way to boost it              |
| Tinker's Complement   | Melter                   | Boostable | COMPLETED |                                                                            |
| Tinker's Construct    | Seared Furnace           | Boostable | COMPLETED | Can only be boosted from controller directly (1 accessible face)           |
| Tinker's Construct    | Smeltery Controller      | Boostable | COMPLETED | Can only be boosted from controller directly                               |
## Files ##
If you want to check the newest version I'll attach the files here. Github has not been updated yet so the sources are also included.
[magicultureintegrations-1.12.2-2.0.0.zip](https://github.com/user-attachments/files/17056266/magicultureintegrations-1.12.2-2.0.0.zip)

**Comment by Krutoy242** [comment]
Glad to hear good news!

Small note: Tinker's Construct `Seared Furnace` could be boosted from 2 faces 


![2020-01-23_16 49 59](https://github.com/user-attachments/assets/3a87c611-80be-4359-b273-f6fdccc96e87)

**Comment by Invadermonky** [comment]
Forgot to mention, the Kekimurus integration from #3 is also included in the version I posted.

**Comment by Invadermonky** [comment]
[Version 1.12.2-2.0.0](https://www.curseforge.com/minecraft/mc-mods/magiculture-integrations/files/5807502) released. Listed features in above [comment](https://github.com/Invadermonky/MagicultureIntegrations/issues/2#issuecomment-2360404008) have been added.

**Comment by Krutoy242** [comment]
I will test all those features later, thank you!

**Comment by Krutoy242** [comment]
Tested many features and seems like they all working fine!
